pandas/core/algorithms.py

```diff
@@ -911,10 +911,15 @@ def select_n_frame(frame, columns, n, method, keep):
     if not is_list_like(columns):
         columns = [columns]
     columns = list(columns)
+    tmp = Series(frame.index)
+    frame.reset_index(inplace=True, drop=True)
     ser = getattr(frame[columns[0]], method)(n, keep=keep)
     if isinstance(ser, Series):
         ser = ser.to_frame()
```

Comment:
so the right way to fix this is to make the internal routines return an indexer instead (e.g. integers), ``select_n_series`` already has this (just needs to return it).

then this is a simple ``.take()``